### PR TITLE
Add unit tests for OAuth2 class

### PR DESCRIPTION
**Pull request by Zain-Abbas1**
- Move buildSessionCookie() into SessionFactory

__JIRA Ticket Number__

- FA-752

### Type of change

* [ ] Bug fix
* [ ] New feature
* [ ] Breaking change

### Pull request checklist

* [ ] I have added unit tests for new code that I have added
* [ ] I have added/updated functional tests where appropriate
* [ ] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__

### CODE REVIEW

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java [hunk 1]
```diff
@@ -45,14 +52,18 @@
...
+    private static UserProfileResponse userProfile;
     private final SignInInfo signInInfo = new SignInInfo();
+    public static MockWebServer mockServer;
     static final String ORIGINAL_REQUEST_URL = "https://www.example.com?original";
-    static final StringBuffer REQUEST_URL_STRING_BUFFER = new StringBuffer(
-            "https://www.example.com"); //TODO decide if we should be expecting StringBuffer here or not
+    static final StringBuffer REQUEST_URL_STRING_BUFFER =
+            new StringBuffer("https://www.example.com"); // TODO decide if we should be expecting
+                                                         // StringBuffer here or not
     public static final String AUTHORISE_URI = "https://example.com/authorise";
```
Comment (CH-JonathanBlake) [inline src/test/java/uk/gov/ch/oauth/Oauth2Test.java:66]: Unused variable.

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java [hunk 2]
```diff
@@ -61,9 +72,21 @@
...
+    @Mock
     private SessionFactory sessionFactory;
     @Mock
     private HttpServletRequest request;
+    @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+    private OAuth2StateHandler oAuth2StateHandler;
+    @Mock
+    private Payload payload;
+    @Mock
+    private JSONObject jsonObject;
```
Comment (CH-JonathanBlake) [inline src/test/java/uk/gov/ch/oauth/Oauth2Test.java:85]: Payload and JSONObject are unused.

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java [hunk 3]
```diff
@@ -45,14 +52,18 @@
...
 public class Oauth2Test {
 
-    public static MockWebServer mockServer;
+    private static final String CODE = "code";
+    private static final String NONCE = "noncevalue";
+    private static final String STATE = "state";
     private static OAuthToken oauthToken;
+    private static UserProfileResponse userProfile;
     private final SignInInfo signInInfo = new SignInInfo();
+    public static MockWebServer mockServer;
```
Comment (CH-JonathanBlake) [inline src/test/java/uk/gov/ch/oauth/Oauth2Test.java:61]: Why are these non constants static?

Comment (CH-JonathanBlake) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:61]: Why is SignIn initiated but UserProfile and Auth token aren't?

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java [hunk 4]
```diff
@@ -61,9 +72,21 @@
...
+    @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+    private OAuth2StateHandler oAuth2StateHandler;
+    @Mock
+    private Payload payload;
+    @Mock
+    private JSONObject jsonObject;
+    @Mock
+    private HttpServletResponse httpServletResponse;
+    @Spy
+    private Map<String, Object> data;
```
Comment (CH-JonathanBlake) [inline src/test/java/uk/gov/ch/oauth/Oauth2Test.java:89]: Data is a very common variable name in this, could we rename the field level to make it more clear what it does.

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java [hunk 5]
```diff
@@ -191,6 +248,55 @@ public void testPrepareState() {
...
+        assertFalse(oauth2.validate(STATE, CODE, httpServletResponse));
+    }
+
+    @Test
+    @DisplayName("Test validate() when getSessionNonce() throws a NullPointerException")
+    public void testGetSessionNonceThrowsNullPointer() {
+        when(oAuth2StateHandler.oauth2DecodeState(anyString()).toJSONObject().getAsString("nonce"))
+                .thenReturn(NONCE);
+        when(sessionFactory.getSessionDataFromContext()).thenReturn(data);
+        when(data.remove(SessionKeys.NONCE.getKey())).thenThrow(new NullPointerException());
```
Comment (CH-JonathanBlake) [inline src/test/java/uk/gov/ch/oauth/Oauth2Test.java:296]: Is this the only place we've spied the data field for?

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java [hunk 6]
```diff
@@ -45,14 +52,18 @@
...
     private static OAuthToken oauthToken;
+    private static UserProfileResponse userProfile;
     private final SignInInfo signInInfo = new SignInInfo();
+    public static MockWebServer mockServer;
     static final String ORIGINAL_REQUEST_URL = "https://www.example.com?original";
-    static final StringBuffer REQUEST_URL_STRING_BUFFER = new StringBuffer(
-            "https://www.example.com"); //TODO decide if we should be expecting StringBuffer here or not
+    static final StringBuffer REQUEST_URL_STRING_BUFFER =
+            new StringBuffer("https://www.example.com"); // TODO decide if we should be expecting
+                                                         // StringBuffer here or not
```
[the remark this answers is not in the document]
Comment (Zain-Abbas1) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:65]: I think this was caught in the reformat I did of the class, not actually my change.

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java (hunk 5 above)
Comment (Zain-Abbas1) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:296]: Yeah the only place, would you like me to move it here locally?

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java (hunk 3 above)
Comment (Zain-Abbas1) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:61]: I believe `MockWebServer` has to be static, which is why I made `userProfile` & `oauthToken` as they are being initialised within the setup() which is a static method.

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (CH-JonathanBlake) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:65]: My point was regarding the ToDo, if the class doesn't use StringBuilder why don't we just make this a String?

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java (hunk 5 above)
Comment (CH-JonathanBlake) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:296]: Why not just do it as real?

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zain-Abbas1) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:65]: Ah it's being used to stub the value of `request.getRequestURL()` which expects a StringBuffer as a result, looks like we have to keep it.

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java (hunk 2 above)
Comment (Zain-Abbas1) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:85]: Strange that eclipse isn't showing these as being unused.

File: src/test/java/uk/gov/ch/oauth/Oauth2Test.java (hunk 5 above)
Comment (Zain-Abbas1) [reply src/test/java/uk/gov/ch/oauth/Oauth2Test.java:296]: As it won't necessary throw an exception in `getSessionNonce` which is what I want to test for.

**Pull request merged**